### PR TITLE
build(1489): Automate setting of default environment variables in Cloud.gov

### LINKED ISSUE #1489: Automate setting of default environment variables in Cloud.gov
**Description:**
This is a follow on to #1036 where we discovered that the cloud.gov applications do not have consistent environment variables in the tanf-dev space and that `set-backend-vars.sh` does not set more than a couple by default. Some spaces like `sandbox` only had 9 variables while others like `raft` had 20+. Inspection of frontend applications reveal no environment variables are set or used presently so these can be ignored.

**Acceptance Criteria:** 
- [ ] All cloud.gov applications (both tanf-dev and tanf-staging) have a consistent set of variables between backends. 
- [ ] When used by the `deploy-backend.sh` or `deploy-frontend.sh`, these scripts perform as expected.
- [ ] Essential environment variables are included by default in these automated scripts and do not have to manually be added by running these scripts.
- [ ] Environment variables should persist across application rebuilds/restages.
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**

- [x] Gather list of common/default backend variables from `tdrs-backend/docker-compose.yml`.
- [x] Update [`set-backend-env-vars.sh` ](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/scripts/set-backend-env-vars.sh#L38-L42) to include new default list
- [x] Update `set-backend-env-vars.sh` to fetch existing vars before an environment is deleted/recreated and ensure they are re-applied on start -- unsure on how this might conflict with "loaded" env vars from the .env files/local environment
- [ ] Ensure deployments remain functionally the same
- [ ] Remove extraneous environment variables from application spaces
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**

**Supporting Documentation:**

**Open Questions:**
- Which spaces have extraneous/obsolete environment variables?
- What is the core set of environment variables needed?
- Do any applications need these environment variables set to differing values? Yes, staging vs raft.

### PR DESCRIPTION
**Pull request by jorgegonzalez**
## Summary of Changes

This PR eliminates the `set-backend-env-vars.sh` script and incorporates it into the `deploy-backend.sh` script as a env function. Additionally the function has been updated and adds a new set of env vars that were consistently available to Docker but not in the deployed environments. As it stands, this branch deploys successfully and is backwards compatible with the current state of `develop`.

Pull request closes #1489

## How to Test

1. To test this PR, deploy this branch to an environment and ensure the proper env vars are set at the correct time.

## Tabletop

- [x] Ready for Tabletop (Reviewer check this)
- [x] Tabletop scheduled (Assignee check this)
- [x] Tabletop approval 

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] All cloud.gov applications (both tanf-dev and tanf-staging) have a consistent set of variables between backends. 
- [x] When used by the `deploy-backend.sh` or `deploy-frontend.sh`, these scripts perform as expected.
- [x] Essential environment variables are included by default in these automated scripts and do not have to manually be added by running these scripts.
- [x] Environment variables should persist across application rebuilds/restages.
- [x] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Comment by andrew-jameson** [comment]
Will be reviewing as-is and will create a different ticket to explore alternative tools/options.

File: scripts/deploy-backend.sh [hunk 1]
```diff
@@ -36,44 +108,80 @@ update_backend()
...
+if [ -n "$BASE_URL" ]; then
+  # Use Shell Parameter Expansion to replace localhost in the URL
+  BASE_URL="${BASE_URL//http:\/\/localhost:8080/$DEFAULT_ROUTE}"
+elif [ "$CF_SPACE" = "tanf-prod" ]; then
+  # Keep the base url set explicitly for production.
+  BASE_URL="$BASE_URL/v1"
+else
+  # Default to the route formed with the cloud.gov env for the lower environments.
+  BASE_URL="$DEFAULT_ROUTE/v1"
+fi
```
Comment (andrew-jameson) [inline scripts/deploy-backend.sh:152]: This section came from `set-backend-vars.sh`.

File: scripts/deploy-backend.sh [hunk 2]
```diff
@@ -36,44 +108,80 @@ update_backend()
...
+DEFAULT_FRONTEND_ROUTE="${DEFAULT_ROUTE//backend/frontend}"
+if [ -n "$FRONTEND_BASE_URL" ]; then
+  FRONTEND_BASE_URL="${FRONTEND_BASE_URL//http:\/\/localhost:3000/$DEFAULT_FRONTEND_ROUTE}"
+elif [ "$CF_SPACE" = "tanf-prod" ]; then
+  # Keep the base url set explicitly for production.
+  FRONTEND_BASE_URL="$FRONTEND_BASE_URL"
+else
+  # Default to the route formed with the cloud.gov env for the lower environments.
+  FRONTEND_BASE_URL="$DEFAULT_FRONTEND_ROUTE"
+fi
```
Comment (andrew-jameson) [inline scripts/deploy-backend.sh:163]: This section came from `set-backend-vars.sh`.

File: scripts/deploy-backend.sh [hunk 3]
```diff
@@ -1,31 +1,103 @@
...
+    # Intentionally not setting variable if empty
+    if [[ -z "${!var_name}" ]]; then
+        echo "WARNING: Empty value for $var_name"
+        continue
+    fi
+    cf_cmd="cf set-env \"$CGAPPNAME_BACKEND\" $var_name \"${!var_name}\""
+    $cf_cmd
+  done
+
+}
```
Comment (andrew-jameson) [inline scripts/deploy-backend.sh:73]: This is an optimization of the repetitious bottom section of `set-backend-vars.sh`.

File: scripts/deploy-backend.sh [hunk 4]
```diff
@@ -1,31 +1,103 @@
...
+
+set_backend_env_vars()
```
Comment (andrew-jameson) [inline scripts/deploy-backend.sh:75]: I'm not convinced this function is very useful/necessary with that "temporary_deployment" flag vs just calling `set_cf_envs` directly from the other sections of the script. The purpose here is to not blindly loop through each and every var. It feels like the result section could just be moved to after #L70

**Comment by andrew-jameson** [comment]
Instead of trying to accomplish the acceptance criteria of "persist environment variables/values across redeploys/restages" all in bash, I will be creating a follow-on ticket to utilize the correct tool(s) for such a solution.

File: scripts/deploy-backend.sh [hunk 5]
```diff
@@ -1,79 +1,161 @@
...
+  "XMS_CLIENT_SECRET")
+
+  for var_name in ${var_list[@]}; do
+    # Intentionally not setting variable if empty
+    if [[ -z "${!var_name}" ]]; then
+        echo "WARNING: Empty value for $var_name"
+        continue
+    fi
+    cf_cmd="cf set-env $CGAPPNAME_BACKEND $var_name ${!var_name}"
+    echo "Debugging: $cf_cmd"
```
Comment (andrew-jameson) [inline scripts/deploy-backend.sh:65]:
```suggestion
```

File: scripts/deploy-backend.sh [hunk 6]
```diff
@@ -1,79 +1,161 @@
...
+  "CLAMAV_NEEDED"
+  "DJANGO_CONFIGURATION"
+  "DJANGO_SECRET_KEY"
+  "DJANGO_SETTINGS_MODULE"
+  "DJANGO_SU_NAME"
+  "FRONTEND_BASE_URL"
+  "JWT_CERT"
+  "JWT_KEY"
+  "LOGGING_LEVEL"
+  "OIDC_RP_CLIENT_ID"
```
Comment (andrew-jameson) [inline scripts/deploy-backend.sh:54]: TODO: Investigate where this is used and why it's needed if at all

File: scripts/deploy-backend.sh [hunk 7]
```diff
@@ -1,79 +1,161 @@
...
+  "XMS_CLIENT_ID"
+  "XMS_CLIENT_SECRET")
```
Comment (andrew-jameson) [inline scripts/deploy-backend.sh:56]:
```suggestion
  )
```

**Review by riatzukiza — APPROVED** [review]

File: scripts/deploy-backend.sh [hunk 8]
```diff
@@ -1,79 +1,159 @@
...
         # set up JWT key if needed
-        if cf e "$CGHOSTNAME_BACKEND" | grep -q JWT_KEY ; then
+        if cf e "$CGAPPNAME_BACKEND" | grep -q JWT_KEY ; then
             echo jwt cert already created
         else
             generate_jwt_cert
         fi
     fi
-    cf map-route "$CGHOSTNAME_BACKEND" app.cloud.gov --hostname "$CGHOSTNAME_BACKEND"
+    set_cf_envs
```
Comment (ADPennington) [inline scripts/deploy-backend.sh:96]:
```suggestion
```

File: scripts/deploy-backend.sh [hunk 9]
```diff
@@ -1,79 +1,158 @@
...
 
+# Dynamically set DJANGO_CONFIGURATION based on Cloud.gov Space
+DJANGO_SETTINGS_MODULE="tdpservice.settings.cloudgov"
+if [ "$CF_SPACE" = "tanf-prod" ]; then
+  DJANGO_CONFIGURATION="Production"
+elif [ "$CF_SPACE" = "tanf-staging" ]; then
+  DJANGO_CONFIGURATION="Staging"
+else
+  DJANGO_CONFIGURATION="Development"
+fi
```
Comment (ADPennington) [inline scripts/deploy-backend.sh:149]:
```suggestion
    set_cf_envs
```

File: scripts/deploy-backend.sh [hunk 10]
```diff
@@ -1,79 +1,158 @@
...
+    set_cf_envs
```
Comment (ADPennington) [inline scripts/deploy-backend.sh:149]:
```suggestion
set_cf_envs
```

File: scripts/deploy-backend.sh [hunk 11]
```diff
@@ -1,79 +1,158 @@
...
+# Dynamically set DJANGO_CONFIGURATION based on Cloud.gov Space
+DJANGO_SETTINGS_MODULE="tdpservice.settings.cloudgov"
+if [ "$CF_SPACE" = "tanf-prod" ]; then
+  DJANGO_CONFIGURATION="Production"
+elif [ "$CF_SPACE" = "tanf-staging" ]; then
+  DJANGO_CONFIGURATION="Staging"
+else
+  DJANGO_CONFIGURATION="Development"
+fi
+set_cf_envs
```
Comment (ADPennington) [inline scripts/deploy-backend.sh:149]:
```suggestion
```

File: scripts/deploy-backend.sh [hunk 12]
```diff
@@ -1,79 +1,157 @@
...
 
-
-    cf bind-service "$CGHOSTNAME_BACKEND" "tdp-staticfiles-${env}"
-    cf bind-service "$CGHOSTNAME_BACKEND" "tdp-datafiles-${env}"
-    cf bind-service "$CGHOSTNAME_BACKEND" "tdp-db-${env}"
-
-    bash ./scripts/set-backend-env-vars.sh "$CGHOSTNAME_BACKEND" "$CF_SPACE"
-
-    cf restage "$CGHOSTNAME_BACKEND"
+    cf restage "$CGAPPNAME_BACKEND"
```
Comment (ADPennington) [inline scripts/deploy-backend.sh:105]:
```suggestion
    set_cf_envs
    cf restage "$CGAPPNAME_BACKEND"
```

File: scripts/deploy-backend.sh [hunk 13]
```diff
@@ -1,79 +1,158 @@
...
 
-
-    cf bind-service "$CGHOSTNAME_BACKEND" "tdp-staticfiles-${env}"
-    cf bind-service "$CGHOSTNAME_BACKEND" "tdp-datafiles-${env}"
-    cf bind-service "$CGHOSTNAME_BACKEND" "tdp-db-${env}"
-
-    bash ./scripts/set-backend-env-vars.sh "$CGHOSTNAME_BACKEND" "$CF_SPACE"
-
-    cf restage "$CGHOSTNAME_BACKEND"
+    set_cf_envs
```
Comment (ADPennington) [inline scripts/deploy-backend.sh:105]:
```suggestion
```

**Comment by andrew-jameson** [comment]
> @andrew-jameson i tested this in the qasp environment by first deleting the apps. Please note:
> 
> * deploy via label with deploystrategy = initial was not successful. i tried moving the set env function around to match closer to the original order in which it was set, but that didnt resolve this.  traceback:
> 
> ```
> 09:44:24.402: [APP/PROC/WEB.0] DJANGO_SUPERUSER_NAME = get_required_env_var_setting(
> 09:44:24.402: [APP/PROC/WEB.0] File "/home/vcap/app/tdpservice/settings/common.py", line 22, in get_required_env_var_setting
> 09:44:24.402: [APP/PROC/WEB.0] raise ImproperlyConfigured(
> 09:44:24.402: [APP/PROC/WEB.0] django.core.exceptions.ImproperlyConfigured: Missing required setting: DJANGO_SUPERUSER_NAME - must set DJANGO_SU_NAME environment variable
> ```
> 
> * after the above failure, i restarted the workflow which utilized the deploystrategy = rolling, since the app existed, and many of the expected env vars were set this time, but you'll notice [here](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/8876/workflows/aea3cdff-fa5a-4394-9f6a-0afbd385bf04/jobs/21950/parallel-runs/0/steps/0-106) that its still missing some (e.g. `jwt_cert`, `av_scan_url`).  this seems consistent with your successful rolling deploy [here](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/8757/workflows/4e34aaa8-0f97-4891-9554-b0ec04af8532/jobs/21529)
> * additionally, the rolling deploy still failed with the following traceback:
> 
> ```
> 13:34:44.778: [APP/PROC/WEB.0] class CloudGov(Common):
> 13:34:44.778: [APP/PROC/WEB.0] File "/home/vcap/app/tdpservice/settings/cloudgov.py", line 43, in CloudGov
> 13:34:44.778: [APP/PROC/WEB.0] cloudgov_services['aws-rds'],
> 13:34:44.778: [APP/PROC/WEB.0] KeyError: 'aws-rds'
> ```

Ok, so while we followed-up via Mattermost, I wanted to lay out what I've done for this ticket the past couple business days.

0. Reverted git commits back to my last merge from develop.
1. Deleted qasp environment then initialized it using script manually
2. Rebuilt the qasp environment by running this script manually and ensuring variables were set and services bound correctly (did not manually bind, script handled this)
3. Ran this script/CI flow using label system with the 'rolling' strategy. At present, this is a known limitation of the CircleCI workflow which invokes this script using the GitHub Actions via the "deploy" labels.
4. Found that CircleCI did not have an AV_SCAN_URL env var?? Re-added this to ensure we continue to have this. 
5. The script now issues warnings for empty env variables which it will skip over.

This should be ready again for QASP.

**Review by ADPennington — APPROVED** [review]
lgtm!

**Pull request merged**